Add Zerproc integration

emlove opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new integration for Zerproc bluetooth LED string lights.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
zerproc:
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13388

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/13388

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [x] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

emlove committed:
Add Zerproc integration

homeassistant commented:
Hi @emlove, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35477).

Once you do that we will be able to review and accept this pull request.

Thanks!

emlove committed:
Add __init__.py for tests module

balloob commented:
Welcome back! 🙂

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,194 @@
+
+    entities = []
+    for light in await asyncio.gather(
+        *[hass.async_add_executor_job(connect_light, light) for light in new_lights]
```
Instead of spawning a lot of executor jobs, do it all in 1 job. We only have limited amount of threads. 

I would probably suggest to convert this whole method to be `sync` since it's actually just a bunch of sync methods except for the last line.

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,194 @@
+        self._brightness = None
+        self._available = True
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown)
```
Do this in `async_added_to_hass` in case the user disables the light.

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,194 @@
+        """Device info for this light."""
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "name": self.name,
```
Can you set manufacturer to `Zerproc`  ?

balloob commented on homeassistant/components/zerproc/strings.json:
```diff
@@ -0,0 +1,14 @@
+  "config": {
+    "step": {
+      "confirm": {
+        "description": "Do you want to set up Zerproc lights?"
+      }
+    },
+    "abort": {
+      "single_instance_allowed": "Only a single configuration of Zerproc lights is necessary.",
+      "no_devices_found": "No Zerproc lights found nearby."
```
I've just merged an update to our scaffold script that lets this use references.
```suggestion
        "description": "[%key:common::config_flow::description::confirm_setup%]"
      }
    },
    "abort": {
      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]",
      "no_devices_found": "[%key:common::config_flow::abort::no_devices_found%]"
```

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,194 @@
+    def update(self):
+        """Fetch new state data for this light."""
+        try:
+            state = self._light.get_state()
```
I am assuming that each light has their own IP. In that case set `PARALLEL_UPDATES = 0` on the top of the file and Home Assistant will talk to each light in parallel.

MartinHjelmare commented on tests/components/zerproc/test_config_flow.py:
```diff
@@ -0,0 +1,30 @@
+        "homeassistant.components.zerproc.config_flow.pyzerproc.discover",
+        return_value=[],
+    ):
+        assert await _async_has_devices(hass) is False
```
Please use the core interfaces to start the tests. Here use `hass.config_entries.flow.async_init` and start the user step.

See eg:
https://github.com/home-assistant/core/blob/823a44cd231c866868ef82c0b3d6cf04475a6539/script/scaffold/templates/config_flow/tests/test_config_flow.py#L9-L16

MartinHjelmare reviewed (changes requested):
Nice to see you again!

We now try to write the tests for the integration more like integration tests.

emlove committed:
Update homeassistant/components/zerproc/strings.json

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

emlove committed:
Review updates

emlove commented:
Thanks for the warm welcome and the reviews! :heart: Feels good to be back.

I pushed up an update that addresses the feedback. It's looking a lot better!

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,198 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown)
```
You will want to undo this listener if the entity is removed earlier. You probably also want to add `async_will_remove_from_hass` that will call `self._light.disconnect()`

```suggestion
        self.async_on_remove(
            self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown)
        )
```

balloob reviewed (approved):
🎉 🎉 Ok to merge when final comment addressed

emlove committed:
Update homeassistant/components/zerproc/light.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

emlove committed:
Additional cleanup

balloob commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,202 @@
+            )
+        )
+
+    def on_hass_shutdown(self, event):
```
Don't forget to also call this from async_will_remove_from_hass in case someone removes config entry

MartinHjelmare commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,202 @@
+        # Double-check the light hasn't been added in another thread
+        if light.address not in hass.data[DOMAIN]["light_entities"]:
+            entity = ZerprocLight(light)
+            hass.data[DOMAIN]["light_entities"][light.address] = entity
```
We don't need to store the entity. Just make the container a `set` and keep track of the light addresses.

MartinHjelmare commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,202 @@
+                warned = True
+
+    # Initial discovery
+    hass.async_add_job(discover)
```
`hass.async_add_job` is legacy for general use. Please use `hass.async_create_task` instead.

ctalkington commented on homeassistant/components/zerproc/light.py:
```diff
@@ -0,0 +1,202 @@
+                EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown
+            )
+        )
+
```
```suggestion

    async def async_will_remove_from_hass(self) -> None:
        """Run when entity about to be removed from hass."""
        self._light.disconnect()

```

emlove replied:
I ended up running it in the thread executor since it's potentially blocking IO

ctalkington replied:
i wondered about that as well after i recommended.

emlove committed:
Additional cleanup

MartinHjelmare commented on homeassistant/components/zerproc/light.py:
```diff
@@ -118,6 +115,10 @@ async def async_added_to_hass(self) -> None:
 
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when entity will be removed from hass."""
+        await self.hass.async_add_executor_job(self._light.disconnect())
```
```suggestion
        await self.hass.async_add_executor_job(self._light.disconnect)
```

emlove replied:
Nice catch, thanks! :woman_facepalming:

MartinHjelmare reviewed (commented):
Looks great! Just a small fix.

emlove committed:
Update homeassistant/components/zerproc/light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Nice!
